README.md
# lazy_load_listview

- [lazy_load_listview](#lazy_load_listview)
- [How to use it.](#how-to-use-it)
- [parameters](#parameters)

Language: [English](README.md)

A simple flutter package for displaying list of items in listview with option to lazy fetch items.

LazyLoadListView provides with 2 different types of options to display Listview rows.
LazyLoadListView provides the liberty to the developers to pass their own custom builder to build custom row. The package provides option to lazy fetch data too.

# Technical Notes:
The layout for pages is built using Listview.builder .
Dev have options show either inbuilt package provided ListTile (Rows) or provide their own designed widgets. 



![](https://github.com/TeaTalkInternal/github_assets/blob/master/gifs/lazy_load_listview_1.gif)![](https://github.com/TeaTalkInternal/github_assets/blob/master/gifs/lazy_load_listview_2.gif)![](https://github.com/TeaTalkInternal/github_assets/blob/master/images/lazy_load_listview.png)


#  How to use it.

The usage is very simple, just use the following code for default rendering of Listview. By default regularImageListItemType row is displayed.

To just go with default implementaion of LazyLoadListView, Use the following code. Here the row will be of type regularImageListItemType.

```dart
 LazyLoadListView(
    listItems: _herosList,
 ),
```
or use the following code to explore power of LazyLoadListView
```dart
 LazyLoadListView(
    listItems: _herosList,
    listItemType: LazyLoadListViewItemType.largeImageListItemType,
    onListItemTap: _didTapList,
    onScrollDidReachEnd: _loadMorePhotos,
    shouldEndLoad: _shouldEndLoadingPhotos,
    fourthRowPrefixIcon: Icons.location_pin,
    loadingIndicatorColor: Colors.grey,
    endOfListText: 'No more results available',
 ),
```
or use the following with to provide your own custom row / ListTile widget

```dart
 LazyLoadListView(
   listItems: _herosList,
   listItemType: LazyLoadListViewItemType.customListItemType,
   customListItemWidgetBuilder: (_context, _index) {
     final _hero = _herosList[_index];
     return CustomListTile(hero: _hero);
   },
   onScrollDidReachEnd: _loadMorePhotos,
   shouldEndLoad: _shouldEndLoadingPhotos,
 ),
```
Here CustomListTile is simple widget created in example code, You may create any and pass it here.

# parameters
## Parameters for lazy_load_listview

| parameter                  | description                                                                           | value                                                                                                                                                                               |
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| listItems                        | Array of items data to be rendered as rows in Listview, Use ListItem object to group item details                                                                  |     \[ListItem]()                                                                                                                                                                              |
| shouldEndLoad          | set the value to true to stop lazy loading of data, Like once you reach end of data from your server or so                                           | This is a optional parameter. value can be true/false                                                                                                                                                    |                                                    |
| listItemType          | Type of List row to be generated                                            | This is a optional parameter. Pass the type of ListItem to be genrated. Value can be either largeImageListItemType, regularImageListItemType or customListItemType.                                                                                                                                                    |
| onScrollDidReachEnd          | handler called when list is scrolled to bottom.                                            | This is a optional parameter. Pass a method where you can handle your logic for lazy loading.                                                                                                                                                    |
| fourthRowPrefixIcon          | The fourth row has a optional prefix icon for text. This can be used to represent phone number, location, email etc.                                            | This is a optional parameter. Pass IconData type value.                                                                                                                                                                                                                                  |
| loadingIndicatorColor          | Color for CircularProgressIndicator for image load or list data load                                            | This is a optional parameter. Pass Color type value. Default value is Colors.grey                                                                                                                                                                                                                                                    |
| endOfListText          | Text or String to be shown once fetch reaches its max limit. No more data available after this point                                            | This is a optional parameter. Default value is "No more results available"                                                                                                                                                                |
| onListItemTap          | ListItem or row click handler                                            | This is a optional parameter.  Pass method to be called on click. Index parameter is returned on click                                                                                  |
| customListItemWidgetBuilder          | Custom IndexedWidgetBuilder (Widget Function(BuildContext context, int index)) . Pass your own custom widget, like ListTile, Card or any.                                           | This is a optional parameter.                                                                                                  |

## Attributes/Properties of ListItem

| parameter                  | description                                                                           | default                                                                                                                                                                               |
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| firstRowText          | String to be displayed in first row Text widget                                          | empty string                                                                                                                                                    |
| secondRowText          | String to be displayed in second row Text widget                                            | empty string                                                                                                                                                    |
| thirdRowText          | String to be displayed in third row Text widget                                            | empty string                                                                                                                                                    |
| forthRowText          | String to be displayed in fourth row Text widget                                            | empty string                                                                                                                                                    |
| imagePath          | Image path of image, either local assets folder image or http image path                                            | empty string                                                                                                                                                    |
| isLocalImage          | Value is true if imagePath is local assets folder image, Value is true if imagePath is http image path                                             | true                                                                                                                                                    |

The usage is well described in the example code.

[git repo here](https://github.com/TeaTalkInternal/lazy_load_listview)

Made with ❤ and dedicated with respect to the Saviour  [Sonu Sood](https://twitter.com/SonuSood)


CHANGELOG.md
## [0.0.1] - 11 May 2021.

* A simple flutter package for displaying list of items in listview with 2 different types of rows and option to lazy fetch data.

## [0.0.2] - 12 May 2021.

* Updated pubspec.yaml and README.mdd

## [0.0.3] - 09 July 2021.

* Added Unique Tag for images via Hero Widget

## [0.0.4] - 09 July 2021.

* Added Unique Id for ListItem

lib/lazy_load_listview.dart
library lazy_load_listview;

import 'package:flutter/material.dart';
import 'package:lazy_load_listview/custom_widgets/large_image_list_Item_widget.dart';
import 'package:lazy_load_listview/custom_widgets/regular_list_Item_widget.dart';
import 'package:lazy_load_listview/models/list_item.dart';

//Types of ListItem
enum LazyLoadListViewItemType {
  largeImageListItemType,
  regularImageListItemType,
  customListItemType
}

class LazyLoadListView extends StatefulWidget {
  final List<ListItem> listItems;
  final Function? onListItemTap;
  final IndexedWidgetBuilder? customListItemWidgetBuilder;
  final LazyLoadListViewItemType listItemType;
  final Function? onScrollDidReachEnd;
  final bool shouldEndLoad;
  final IconData? fourthRowPrefixIcon;
  final String endOfListText;
  final Color loadingIndicatorColor;

  LazyLoadListView({
    required this.listItems,
    this.onListItemTap,
    this.onScrollDidReachEnd,
    this.listItemType = LazyLoadListViewItemType.regularImageListItemType,
    this.shouldEndLoad = false,
    this.fourthRowPrefixIcon,
    this.endOfListText = 'No more results available',
    this.loadingIndicatorColor = Colors.grey,
    this.customListItemWidgetBuilder,
  });

  @override
  _LazyLoadListViewState createState() => _LazyLoadListViewState();
}

class _LazyLoadListViewState extends State<LazyLoadListView> {
  ScrollController _scrollController = ScrollController();
  bool _scrollDidReachEnd = false;
  @override
  void initState() {
    super.initState();
    _scrollController.addListener(() {
      if (_scrollController.position.pixels ==
          _scrollController.position.maxScrollExtent) {
        //Detect scroll position here
        setState(() {
          _scrollDidReachEnd = true;
        });
        if (widget.onScrollDidReachEnd != null) {
          widget.onScrollDidReachEnd!();
          return;
        }
      }
    });
  }

  @override
  void dispose() {
    _scrollController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      controller: _scrollController,
      padding: EdgeInsets.only(
        top: 5.0,
        bottom: 30.0,
      ),
      scrollDirection: Axis.vertical,
      itemBuilder: (context, index) {
        return _buildListTileItem(context, index);
      },
      itemCount: _getWidgetsCount(),
    );
  }

  int _getWidgetsCount() {
    if (widget.listItems.length == 0) {
      return 0;
    } else if (widget.onScrollDidReachEnd == null ||
        (widget.onScrollDidReachEnd != null && _scrollDidReachEnd == false)) {
      return widget.listItems.length;
    } else if (widget.shouldEndLoad == true) {
      return widget.listItems.length + 1;
    }
    return widget.listItems.length + 1;
  }

  Widget _buildListTileItem(BuildContext context, int index) {
    if (index == widget.listItems.length && widget.shouldEndLoad == true) {
      return Center(
          child: Padding(
        padding: const EdgeInsets.only(
          top: 15.0,
          bottom: 10.0,
        ),
        child: Text(
          widget.endOfListText,
          style: TextStyle(color: Colors.grey, fontSize: 15.9),
        ),
      ));
    } else if (index == widget.listItems.length &&
        widget.shouldEndLoad == false) {
      return Center(
        child: Padding(
          padding: const EdgeInsets.only(top: 30.0),
          child: CircularProgressIndicator(
            valueColor: new AlwaysStoppedAnimation<Color>(
              widget.loadingIndicatorColor,
            ),
          ),
        ),
      );
    } else {
      // return (widget.listItemType ==
      //         LazyLoadListViewItemType.largeImageListItemType)
      //     ? _buildLargeImageListViewWidget(context, index)
      //     : _buildRegularListViewWidget(context, index);
      if (widget.listItemType ==
          LazyLoadListViewItemType.largeImageListItemType) {
        return _buildLargeImageListViewWidget(context, index);
      } else if (widget.listItemType ==
              LazyLoadListViewItemType.customListItemType &&
          widget.customListItemWidgetBuilder != null) {
        return widget.customListItemWidgetBuilder!(context, index);
      }
      return _buildRegularListViewWidget(context, index);
    }
  }

  //Two variations of List Items.
  //Large Image ListItem
  Widget _buildLargeImageListViewWidget(BuildContext context, int index) {
    final ListItem _listItem = widget.listItems[index];
    return LargeImageListItemWidget(
      onListItemTap: widget.onListItemTap,
      listItem: _listItem,
      listItemIndex: index,
      fourthRowPrefixIcon: widget.fourthRowPrefixIcon,
      loadingIndicatorColor: widget.loadingIndicatorColor,
    );
  }

  //Regular Image ListItem
  Widget _buildRegularListViewWidget(BuildContext context, int index) {
    final ListItem _listItem = widget.listItems[index];
    return RegularListItemWidget(
      onListItemTap: widget.onListItemTap,
      listItem: _listItem,
      listItemIndex: index,
      fourthRowPrefixIcon: widget.fourthRowPrefixIcon,
      loadingIndicatorColor: widget.loadingIndicatorColor,
    );
  }
}


lib/custom_widgets/regular_list_Item_widget.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:lazy_load_listview/models/list_item.dart';

class RegularListItemWidget extends StatelessWidget {
  final Function? onListItemTap;
  final int listItemIndex;
  final ListItem listItem;
  final IconData? fourthRowPrefixIcon;
  final Color loadingIndicatorColor;

  const RegularListItemWidget({
    Key? key,
    required this.onListItemTap,
    required this.listItemIndex,
    required this.listItem,
    this.fourthRowPrefixIcon,
    required this.loadingIndicatorColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        if (onListItemTap != null) {
          onListItemTap!(listItemIndex);
        }
      },
      child: Card(
        margin: EdgeInsets.only(
          top: 15.0,
          left: 5.0,
          right: 5.0,
        ),
        elevation: 7.0,
        shadowColor: Colors.grey[200],
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(20.0),
        ),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            RegularListRowThumbnailImageWidget(
              listItem: listItem,
              loadingIndicatorColor: loadingIndicatorColor,
            ),
            Flexible(
              flex: 4,
              child: RegularListRowDetailsWidget(
                listItem: listItem,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

//Row Description Widget
class RegularListRowDetailsWidget extends StatelessWidget {
  final ListItem listItem;
  final IconData? fourthRowPrefixIcon;

  const RegularListRowDetailsWidget({
    Key? key,
    required this.listItem,
    this.fourthRowPrefixIcon,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.only(
        right: 5.0,
        top: 10.0,
        bottom: 10.0,
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            listItem.firstRowText,
            style: TextStyle(
              color: Colors.black,
              fontSize: 22.0,
              fontWeight: FontWeight.w900,
            ),
            maxLines: 2,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Text(
            listItem.secondRowText,
            style: TextStyle(
              color: Colors.grey[700],
              fontSize: 17.0,
              fontWeight: FontWeight.w600,
            ),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Text(
            listItem.thirdRowText,
            style: TextStyle(
              color: Colors.grey[700],
              fontSize: 15.0,
              fontWeight: FontWeight.w400,
            ),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.start,
            children: [
              if (fourthRowPrefixIcon != null)
                Icon(
                  fourthRowPrefixIcon!,
                  color: Colors.grey[600],
                  size: 17.0,
                ),
              Flexible(
                child: Text(
                  listItem.forthRowText,
                  style: TextStyle(
                    color: Colors.grey[600],
                    fontSize: 14.0,
                    fontWeight: FontWeight.w400,
                  ),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                ),
              ),
            ],
          )
        ],
      ),
    );
  }
}

//Row Image Widget
class RegularListRowThumbnailImageWidget extends StatelessWidget {
  final ListItem listItem;
  final Color loadingIndicatorColor;

  const RegularListRowThumbnailImageWidget({
    Key? key,
    required this.listItem,
    required this.loadingIndicatorColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Hero(
      tag: listItem.listItemId,
      child: Container(
        width: 120.0,
        height: 135.0,
        margin: EdgeInsets.all(12.0),
        clipBehavior: Clip.hardEdge,
        child: listItem.isLocalImage
            ? Image.asset(
                listItem.imagePath,
                fit: BoxFit.cover,
              )
            : CachedNetworkImage(
                imageUrl: listItem.imagePath,
                fit: BoxFit.cover,
                placeholder: (context, url) {
                  return Center(
                    child: SizedBox(
                      height: 30.0,
                      width: 30.0,
                      child: CircularProgressIndicator(
                          valueColor: new AlwaysStoppedAnimation<Color>(
                        loadingIndicatorColor,
                      )),
                    ),
                  );
                },
              ),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(20.0),
        ),
      ),
    );
  }
}


lib/custom_widgets/large_image_list_Item_widget.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:lazy_load_listview/models/list_item.dart';

class LargeImageListItemWidget extends StatelessWidget {
  final Function? onListItemTap;
  final int listItemIndex;
  final ListItem listItem;
  final IconData? fourthRowPrefixIcon;
  final Color loadingIndicatorColor;

  const LargeImageListItemWidget({
    Key? key,
    required this.onListItemTap,
    required this.listItemIndex,
    required this.listItem,
    this.fourthRowPrefixIcon,
    required this.loadingIndicatorColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: () {
        if (onListItemTap != null) {
          onListItemTap!(listItemIndex);
        }
      },
      child: Container(
        padding: EdgeInsets.only(
          bottom: 15.0,
          left: 0.0,
        ),
        height: 220,
        child: Row(
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            Flexible(
              flex: 3,
              child: LargeImageListRowThumbnailImageWidget(
                listItem: listItem,
                loadingIndicatorColor: loadingIndicatorColor,
              ),
            ),
            Flexible(
              flex: 4,
              child: LargeImageListRowDetailsWidget(
                listItem: listItem,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

//Row Description Widget
class LargeImageListRowDetailsWidget extends StatelessWidget {
  final ListItem listItem;
  final IconData? fourthRowPrefixIcon;

  static final BoxShadow _cardShadow = BoxShadow(
    offset: Offset(2, 2),
    color: Colors.grey[300] ?? Colors.grey,
    blurRadius: 7.0,
    spreadRadius: 3.0,
  );

  const LargeImageListRowDetailsWidget({
    Key? key,
    required this.listItem,
    this.fourthRowPrefixIcon,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: EdgeInsets.only(
        right: 3.0,
        top: 20.0,
        bottom: 20.0,
      ),
      padding: EdgeInsets.symmetric(horizontal: 10.0, vertical: 10.0),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.only(
          topRight: Radius.circular(15.0),
          bottomRight: Radius.circular(15.0),
        ),
        boxShadow: [_cardShadow],
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Text(
            listItem.firstRowText,
            style: TextStyle(
              color: Colors.black,
              fontSize: 22.0,
              fontWeight: FontWeight.w900,
            ),
            maxLines: 2,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Text(
            listItem.secondRowText,
            style: TextStyle(
              color: Colors.grey[700],
              fontSize: 17.0,
              fontWeight: FontWeight.w600,
            ),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Text(
            listItem.thirdRowText,
            style: TextStyle(
              color: Colors.grey[700],
              fontSize: 15.0,
              fontWeight: FontWeight.w400,
            ),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
          SizedBox(
            height: 8.0,
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.start,
            children: [
              if (fourthRowPrefixIcon != null)
                Icon(
                  fourthRowPrefixIcon!,
                  color: Colors.grey[600],
                  size: 17.0,
                ),
              Flexible(
                child: Text(
                  listItem.forthRowText,
                  style: TextStyle(
                    color: Colors.grey[600],
                    fontSize: 14.0,
                    fontWeight: FontWeight.w400,
                  ),
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                ),
              ),
            ],
          )
        ],
      ),
    );
  }
}

//Image widget
class LargeImageListRowThumbnailImageWidget extends StatelessWidget {
  final ListItem listItem;
  final Color loadingIndicatorColor;

  const LargeImageListRowThumbnailImageWidget({
    Key? key,
    required this.listItem,
    required this.loadingIndicatorColor,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Hero(
      tag: listItem.listItemId,
      child: Container(
        height: 220,
        width: double.infinity,
        clipBehavior: Clip.hardEdge,
        child: listItem.isLocalImage
            ? Image.asset(
                listItem.imagePath,
                fit: BoxFit.cover,
              )
            : CachedNetworkImage(
                imageUrl: listItem.imagePath,
                fit: BoxFit.cover,
                placeholder: (context, url) {
                  return Center(
                    child: SizedBox(
                      height: 30.0,
                      width: 30.0,
                      child: CircularProgressIndicator(
                          valueColor: new AlwaysStoppedAnimation<Color>(
                        loadingIndicatorColor,
                      )),
                    ),
                  );
                },
              ),
        decoration: BoxDecoration(
          borderRadius: BorderRadius.circular(20.0),
        ),
      ),
    );
  }
}


example/lib/example_widget.dart
import 'package:flutter/material.dart';
import 'package:lazy_load_listview/lazy_load_listview.dart';
import 'package:lazy_load_listview/models/list_item.dart';

class ExampleWidget extends StatefulWidget {
  @override
  _ExampleWidgetState createState() => _ExampleWidgetState();
}

class _ExampleWidgetState extends State<ExampleWidget> {
  List<ListItem> _herosList = [
    ListItem(
      firstRowText: 'Sonu Sood',
      secondRowText: 'Actor',
      thirdRowText: 'Man with million hearts',
      forthRowText: 'Mumbai, Maharashtra, India',
      imagePath:
          'https://images.indianexpress.com/2021/05/sonu-sood-1200-1.jpg',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Dr. Shankare Gowda',
      secondRowText: 'Doctor',
      thirdRowText: 'A skin specialist who treats patients for Rs 5',
      forthRowText: 'Mandya, Karnataka, India',
      imagePath:
          'https://new-img.patrika.com/upload/2018/04/24/five_rupees_doctor_2700143_835x547-m.jpg',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Babar Ali',
      secondRowText: 'Youngest Headmaster',
      thirdRowText: 'Teacher for poor',
      forthRowText: 'Murshidabad, West Bengal, India',
      imagePath:
          'https://upload.wikimedia.org/wikipedia/commons/e/eb/Babar-Ali-Portfolio-Image.jpg',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Anand Kumar',
      secondRowText: 'Mathematics Educator',
      thirdRowText: 'Coaches underprivileged students',
      forthRowText: 'Patna, Bihar, India',
      imagePath:
          'https://www.jagranjosh.com/imported/images/E/GK/anand-kumar.jpg',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Shivakumara Swami',
      secondRowText: 'Humanitarian, spiritual leader & educator',
      thirdRowText: 'He founded a total of 132 institutions for education',
      forthRowText: 'Tumkur, Karnataka, India',
      imagePath:
          'https://www.filmibeat.com/img/popcorn/profile_photos/sri-shivakumara-swamy-20190121111407-7590.jpg',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Kareem Bhai',
      secondRowText: 'Retired daily-wage laborer',
      thirdRowText: 'Distributes free newspaper for children',
      forthRowText: 'Abids, Hyderabad, India',
      imagePath:
          'https://www.theink24.com/wp-content/uploads/2018/06/paperboy_kareem-bhai.png',
      isLocalImage: false,
    ),
    ListItem(
      firstRowText: 'Salumarada Thimmakka',
      secondRowText: 'Environmentalist',
      thirdRowText: 'Planted more than 385 trees to help travellers',
      forthRowText: 'Gubbi Taluk, Tumkur, India',
      imagePath:
          'https://www.newskarnataka.com/storage/photos/shares/2018MAY/saalumarada_thimmakka.jpg',
      isLocalImage: false,
    ),
  ];

  bool _shouldEndLoadingPhotos = false;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        padding: EdgeInsets.symmetric(
          horizontal: 15.0,
        ),
        color: Colors.grey[100],
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            DiscoverAppBarWidget(),
            //To explore all the Customisation options to load the listview use the following commented code

            Expanded(
              child: LazyLoadListView(
                listItems: _herosList,
                listItemType: LazyLoadListViewItemType.largeImageListItemType,
                onListItemTap: _didTapList,
                onScrollDidReachEnd: _loadMorePhotos,
                shouldEndLoad: _shouldEndLoadingPhotos,
                fourthRowPrefixIcon: Icons.location_pin,
                loadingIndicatorColor: Colors.grey,
                endOfListText: 'No more results available',
              ),
            ),
            /*
            //To simply load list use the following commented code
            Expanded(
              child: LazyLoadListView(
                listItems: _herosList,
              ),
            ),
          */
            //To simply use your own custom list item list use the following commented code
            /*
            Expanded(
              child: LazyLoadListView(
                listItems: _herosList,
                listItemType: LazyLoadListViewItemType.customListItemType,
                customListItemWidgetBuilder: (_context, _index) {
                  final _hero = _herosList[_index];
                  return CustomListTile(hero: _hero);
                },
                onScrollDidReachEnd: _loadMorePhotos,
                shouldEndLoad: _shouldEndLoadingPhotos,
              ),
            ),
          */
          ],
        ),
      ),
    );
  }

//Callbacks
  void _didTapList(int _index) {
    print('Index $_index');
  }

  void _loadMorePhotos() {
    //You may values fetched from server and append to your datasource.
    //here we have added a future delay to mock a server call
    if (_herosList.length < 8) {
      Future.delayed(Duration(seconds: 1), () {
        setState(() {
          _herosList.addAll([
            ListItem(
              firstRowText: 'Omkarnath Sharma',
              secondRowText: 'Retired blood bank technician',
              thirdRowText:
                  'Has set free medicine bank for the underprivileged and poor',
              forthRowText: 'Delhi, India',
              imagePath:
                  'https://nkhayi.files.wordpress.com/2012/05/009-omkar-nath.jpg',
              isLocalImage: false,
            ),
            ListItem(
              firstRowText: 'Mamta Rawat',
              secondRowText: 'Part-time trainer for the NIM',
              thirdRowText:
                  'Saved thousands of lives in the Uttarakhand flood of 2013',
              forthRowText: 'Uttarakhand, Bankoli, India',
              imagePath: 'https://i.ytimg.com/vi/o9rkNiA-awI/maxresdefault.jpg',
              isLocalImage: false,
            ),
            ListItem(
              firstRowText: 'Dr APJ Abdul Kalam',
              secondRowText: 'Aerospace Scientist, Author',
              thirdRowText: 'An inspiration to many',
              forthRowText: 'Rameswaram, Tamil Nadu, India',
              imagePath:
                  'https://www.apjabdulkalamfoundation.org/images/about/abdulkalamprofile.jpg',
              isLocalImage: false,
            ),
          ]);
        });
      });
    } else {
      setState(() {
        _shouldEndLoadingPhotos = true;
      });
    }
  }
}

//Sub-Widgets
//Header
class DiscoverAppBarWidget extends StatelessWidget {
  const DiscoverAppBarWidget({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.only(
        top: 60.0,
        left: 5.0,
        bottom: 10.0,
      ),
      child: Text(
        'Real life Heros',
        style: TextStyle(
            fontWeight: FontWeight.w900,
            fontSize: 32.0,
            color: Colors.grey[850]),
      ),
    );
  }
}

//Custom List Item
class CustomListTile extends StatelessWidget {
  const CustomListTile({
    Key? key,
    required ListItem hero,
  })   : _hero = hero,
        super(key: key);

  final ListItem _hero;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        ListTile(
          leading: CircleAvatar(
            backgroundImage: NetworkImage(
              _hero.imagePath,
            ),
            radius: 40.0,
          ),
          contentPadding: EdgeInsets.all(0.0),
          title: Text(
            _hero.firstRowText,
            style: TextStyle(
              color: Colors.black,
              fontSize: 17.0,
              fontWeight: FontWeight.w900,
            ),
            maxLines: 2,
            overflow: TextOverflow.ellipsis,
          ),
          subtitle: Text(
            _hero.secondRowText,
            style: TextStyle(
              color: Colors.grey[700],
              fontSize: 14.0,
              fontWeight: FontWeight.w400,
            ),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
          ),
          trailing: Icon(
            Icons.arrow_forward_ios,
            size: 15.0,
            color: Colors.grey,
          ),
        ),
        Divider(
          indent: 75.0,
          color: Colors.grey,
        )
      ],
    );
  }
}
